### PR TITLE
feat(gateway)!: validate registry init args against ABI (ENG-463)

### PR DESCRIPTION
**Pull request by peer2f00l**
## Summary
- resolve registry WASM by trusted hash across catalogued, global, and stored versions
- validate JSON init args against the embedded constructor ABI before deployment planning
- replace version-key constructor heuristics with shared ABI validation and explicit opt-out

## Verification
- cargo check --workspace
- cargo test -p templar-gateway-methods-dispatch --lib -- --nocapture
- cargo test -p templar-manager -- --nocapture
- cargo test -p templar-gateway-core --lib -- --nocapture
- cargo test -p templar-gateway-catalog methods_md_is_up_to_date
- just test-sandbox -p templar-gateway-service --test-threads=1 --stale
- cargo fmt --all --check
- affected-package cargo clippy --all-targets -- -D warnings

- - -
This change is [Reviewable](https://reviewable.io/reviews/Templar-Protocol/contracts/607)

### CODE REVIEW

File: gateway/methods-dispatch/src/contract_abi.rs [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+                };
+                if let Ok(abi) = serde_json::from_slice(&decoded) {
+                    return Ok(abi);
+                }
+            }
+        }
+    }
+
+    Err(abi_error("registry deployment WASM has no embedded ABI"))
+}
```
[the remark this answers is not in the document]
Comment (peer2f00l) [reply gateway/methods-dispatch/src/contract_abi.rs:120]: Implemented. ABI candidate probing is capped at 16 candidates across all Wasm data sections, with an explicit cap error. The complete synchronous parse/decompress/schema-validation path now runs in spawn_blocking; worker join failures use GatewayError::Internal. Dispatch tests cover the cap and the blocking boundary’s result path.

File: gateway/methods-dispatch/src/contract_abi.rs [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+        let parameter_schema = serde_json::to_value(&parameter.type_schema).map_err(|error| {
+            abi_error(format!(
+                "registry deployment constructor ABI cannot be encoded as JSON Schema: {error}"
+            ))
+        })?;
+        if !allows_null(&parameter_schema) {
+            required.push(Value::String(parameter.name.clone()));
+        }
+        properties.insert(parameter.name.clone(), parameter_schema);
+    }
```
[the remark this answers is not in the document]
Comment (peer2f00l) [reply gateway/methods-dispatch/src/contract_abi.rs:164]: Implemented structurally. allows_null was replaced with Draft 7 validation of a wrapper schema containing the ABI definitions and allOf parameter schema. Top-level defaults, nullable type arrays, nested unions, and definition-backed nullable references are covered by tests.

File: gateway/methods-dispatch/src/contract_abi.rs [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+            .filter_map(|key| schema.get(key).and_then(Value::as_array))
+            .flatten()
+            .any(|schema| schema.get("type").and_then(Value::as_str) == Some("null"))
+}
+
+fn abi_error(reason: impl std::fmt::Display) -> GatewayError {
+    GatewayError::RequestPreconditionFailed(format!(
+        "{reason}; pass --skip-abi-check to bypass ABI validation"
+    ))
+}
```
[the remark this answers is not in the document]
Comment (peer2f00l) [reply gateway/methods-dispatch/src/contract_abi.rs:211]: Implemented. The bypass hint now uses the serialized gateway field name: set skip_abi_check=true to bypass ABI validation.

File: gateway/methods-dispatch/src/contract_abi.rs [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+        let error = validate_constructor_args(&wasm, br"{}")
+            .expect_err("missing required constructor argument");
+        assert!(error.to_string().contains("do not match"));
+        let error = validate_constructor_args(&wasm, br#"{"owner_id":3}"#)
+            .expect_err("wrong constructor argument type");
+        assert!(error.to_string().contains("do not match"));
+        let error = validate_constructor_args(&wasm, br#"{"owner_id":"owner.near","extra":true}"#)
+            .expect_err("unknown constructor argument");
+        assert!(error.to_string().contains("do not match"));
+    }
```
[the remark this answers is not in the document]
Comment (peer2f00l) [reply gateway/methods-dispatch/src/contract_abi.rs:287]: Implemented. The valid constructor case is separate, while missing, wrong-type, and extra-key failures are parameterized with rstest; focused tests cover extraction and optionality boundaries.

File: gateway/methods-dispatch/src/registry_impl.rs [hunk 5]
```diff
@@ -212,6 +214,17 @@ pub(crate) async fn plan_create_from_registry<C: HasNearClient>(
...
+    if !target.skip_abi_check {
+        let wasm = resolve_registry_wasm(
+            ctx,
+            &target.registry_id,
+            registry_version,
+            &target.version_key,
+        )
+        .await?;
+        validate_constructor_args(&wasm, &init_args)?;
+    }
```
[the remark this answers is not in the document]
Comment (peer2f00l) [reply gateway/methods-dispatch/src/registry_impl.rs:226]: Reviewed without a code change. This is a planning input used by in-process, CLI, relayer-configured, and RPC consumers; transaction authority remains the managed signer boundary. Adding a method-specific authorization check would not secure the otherwise equivalent write API.

File: gateway/methods-dispatch/src/registry_impl.rs [hunk 6]
```diff
@@ -212,6 +214,17 @@ pub(crate) async fn plan_create_from_registry<C: HasNearClient>(
...
         .cached_version()
         .await?;
+    if !target.skip_abi_check {
+        let wasm = resolve_registry_wasm(
+            ctx,
+            &target.registry_id,
+            registry_version,
+            &target.version_key,
+        )
+        .await?;
```
[the remark this answers is not in the document]
Comment (peer2f00l) [reply gateway/methods-dispatch/src/registry_impl.rs:224]: Reviewed without a code change. Availability and exact hash checks remain per-plan. A global cache would add mutable state and eviction semantics without a measured correctness or performance requirement.

File: gateway/methods-dispatch/src/registry_wasm.rs [hunk 7]
```diff
@@ -0,0 +1,263 @@
...
+            OnchainCodeSource::Stored(StoredCodeLen(1)),
+            CryptoHash([0; 32]),
+            None,
+        )
+        .await
+        .expect_err("an unavailable catalog must continue to the onchain source");
+
+        assert!(matches!(error, GatewayError::NearQuery(_)));
+    }
+}
```
[the remark this answers is not in the document]
Comment (peer2f00l) [reply gateway/methods-dispatch/src/registry_wasm.rs:406]: Implemented. StoredCodeAssembler owns offset and allocation state, validates missing, empty, oversized, and overrun chunks before mutation, and tests multi-chunk progression and final assembly.

File: gateway/methods-spec/src/registry.rs [hunk 8]
```diff
@@ -105,20 +105,24 @@ pub struct AddVersion {
...
+/// Unmodeled contracts use [`Deploy`]; its JSON init args are validated against the
+/// target constructor ABI unless `skip_abi_check` is set.
 #[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
 pub struct DeployTarget {
     pub registry_id: AccountId,
     pub name: String,
     pub version_key: String,
+    /// Skip embedded-ABI validation for a deployment whose constructor cannot be checked.
+    #[serde(default, skip_serializing_if = "std::ops::Not::not")]
+    pub skip_abi_check: bool,
```
[the remark this answers is not in the document]
Comment (peer2f00l) [reply gateway/methods-spec/src/registry.rs:118]: Implemented through repository policy. The PR title is now feat(gateway)!: validate registry init args against ABI; Cargo versions remain release-plz managed.

File: service/gateway/src/rpc/tests/registry_tests.rs [hunk 9]
```diff
@@ -224,3 +227,197 @@ async fn add_version_accepts_every_source_and_each_one_deploys() -> Result<()> {
...
+        .request::<registry::ListDeployments>(&registry::ListDeployments {
+            registry_id,
+            args: Pagination::default(),
+        })
+        .await?;
+    assert!(deployments.account_ids.is_empty());
+
+    stack.shutdown().await;
+    Ok(())
+}
```
[the remark this answers is not in the document]
Comment (peer2f00l) [reply service/gateway/src/rpc/tests/registry_tests.rs:392]: Implemented. The sandbox fixture is now a deployable Wasm module with Type, Function, Memory, Export, Code, and Data sections. The acceptance test validates rejected unexpected args and successful empty args for both Stored and PublishGlobal versions, including succeeded operations and final deployment records.

File: gateway/methods-dispatch/src/registry_wasm.rs [hunk 10]
```diff
@@ -0,0 +1,405 @@
...
+        let chunk = client.get_version_code_chunk(&request).await?;
+        if chunk.is_none() {
+            return Err(precondition(format!(
+                "registry version {version_key} no longer has stored code"
+            )));
+        }
+        assembler.push(version_key, len, chunk)?;
+    }
+
+    assembler.finish()
```
[the remark this answers is not in the document]
Comment (peer2f00l) [reply gateway/methods-dispatch/src/registry_wasm.rs:244]: Addressed in commit 497f7476. StoredCodeAssembler::push now rejects any response whose length differs from requested_len before offset or bytes mutate. Since read_stored_code requests min(remaining, CODE_CHUNK_LEN), successful reads are bounded to ceil(code_len / CODE_CHUNK_LEN) calls. Added regression coverage for partial responses and verified the targeted/full dispatch tests, sandbox smoke test, fast gate, formatting, Clippy, and workspace checks.

**Comment by peer2f00l** [comment]
@coderabbitai review

**Comment by peer2f00l** [comment]
@codex review

File: gateway/methods-dispatch/src/contract_abi.rs [hunk 11]
```diff
@@ -0,0 +1,369 @@
...
+        .enumerate()
+        .filter_map(|(offset, window)| (window == ZSTD_MAGIC).then_some(offset))
+}
+
+fn decompress(compressed: &[u8]) -> std::io::Result<Vec<u8>> {
+    let zstd_decoder = zstd::stream::read::Decoder::new(Cursor::new(compressed))?;
+    let mut decoded = Vec::new();
+    zstd_decoder
+        .take(MAX_ABI_BYTES + 1)
+        .read_to_end(&mut decoded)?;
```
[the remark this answers is not in the document]
Comment (peer2f00l) [reply gateway/methods-dispatch/src/contract_abi.rs]: Fixed in 9dc0998b. `decompress` now configures the streaming zstd decoder with `single_frame()`, so unrelated bytes after the embedded ABI frame are ignored while the existing bounded read, 8 MiB limit, and malformed-frame diagnostics remain unchanged. Added `accepts_embedded_abi_before_trailing_segment_data` to cover the real artifact layout. Focused and full methods-dispatch tests, sandbox registry validation, fast tests, Clippy, formatting, catalog tests, and workspace check passed.
